src/audio/pipewire.c

```diff
@@ -0,0 +1,322 @@
+    state->sample_buffer = malloc(SAMPLE_BUFFER_SIZE);
+    // we don't know if the params array belongs to us
+    // it  could be freed by speech dispatcher at a later date, which means we could be storing a pointer to freed memory
+    // to ensure that the params array belongs to us and only we can free it, we will duplicate the string contained in params[1], even if such an operation would have not been necesary in the end
+    state->sink_name = strdup(pars[1]);
+    // initialise the event file descripter and the stream
+    pw_thread_loop_lock(state->loop);
+    state->stream = pw_stream_new_simple(pw_thread_loop_get_loop(state->loop), state->sink_name, pw_properties_new(PW_KEY_MEDIA_TYPE, "Audio", PW_KEY_MEDIA_CATEGORY, "Playback", PW_KEY_MEDIA_ROLE, "Accessibility", NULL), &stream_events, state);
+    if ((state->eventfd_number = spa_system_eventfd_create(pw_thread_loop_get_loop(state->loop)->system, SPA_FD_CLOEXEC)) < 0)
+        return NULL;
+    pw_thread_loop_unlock(state->loop);
+    // start the threaded loop
+    pw_thread_loop_start(state->loop);
+    return (AudioID *)state;
+}
+// configure a pipewire stream for the given speech dispatcher configuration and then prepair loop for playback
+static int pipewire_begin(AudioID *id, AudioTrack track)
+{
+    module_state *state = (module_state *)id;
+    const struct spa_pod *params;
+    uint8_t buffer[1024];
+    struct spa_pod_builder b = SPA_POD_BUILDER_INIT(buffer, sizeof(buffer));
+    int format;
+    if (track.bits == 16)
+    {
+        state->stride = track.num_channels * sizeof(uint16_t);
+        switch (state->id.format)
+        {
+        case SPD_AUDIO_LE:
+            format = SPA_AUDIO_FORMAT_S16_LE;
+            break;
+        case SPD_AUDIO_BE:
+            format = SPA_AUDIO_FORMAT_S16_BE;
+            break;
+        default:
+            pw_log_error("invalid audio format specifier");
+            return -1;
+        }
+    }
+    else if (track.bits == 8)
+    {
+        state->stride = track.num_channels * sizeof(uint8_t);
+        format = SPA_AUDIO_FORMAT_S8;
+    }
+    else
+    {
+        pw_log_error("invalid audio format specifier");
+        return -1;
+    }
+
+    params = spa_format_audio_raw_build(&b, SPA_PARAM_EnumFormat, &SPA_AUDIO_INFO_RAW_INIT(.format = format, .channels = track.num_channels, .rate = track.sample_rate));
+    pw_thread_loop_lock(state->loop);
+    pw_stream_connect(state->stream, PW_DIRECTION_OUTPUT, PW_ID_ANY, PW_STREAM_FLAG_AUTOCONNECT | PW_STREAM_FLAG_MAP_BUFFERS | PW_STREAM_FLAG_RT_PROCESS, &params, 1);
+    pw_thread_loop_unlock(state->loop);
+    return 0;
+}
+static int pipewire_feed_sink_overlap(AudioID *id, AudioTrack track)
+{
+    module_state *state = (module_state *)id;
+    uint32_t write_index, fill_quantity, room_left_in_buffer = 0, track_buffer_size;
+    uint64_t dummy; // used to read from the event file descripter. We don't use this for anything, at the moment, it's always 42
```

Comment:
Actually it will be more than 42 if `on_process` is called several times before `pipewire_feed_sink_overlap` gets to execute.

Reply:
It is sort of a dummy value, yes, but it still adds up each time on_process is called, and will thus be a multiple of 42.